### ISSUE TITLE
Make sure that any content type permissions are removed from roles when the respective content type is deleted

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5350]
## Steps To Reproduce
1. create a new content type
2. on the "Permissions" vertical tab, assign some permissions to the "Editor" role
3. save the content type
4. navigate to `admin/config/development/configuration/single/export`
5. select "User roles" in the "Configuration group" drop-down
6. select "Administrator" in the "Configuration name" drop-down
7. scroll to the end of the permissions list shown in the "Exported configuration" field -> verify that the permissions for the new content type are there 👍🏼 
8. switch to the "Editor" role via the "Configuration name" drop-down
9. scroll to the end of the permissions list shown in the "Exported configuration" field -> verify that the permissions for the new content type are there 👍🏼 
10. delete the content type you have created for the purposes of this test
11. navigate to `admin/config/development/configuration/single/export` and repeat steps 5 through 9 above...

## Actual behavior

The permissions for the content type that was just deleted are still there for all the roles they were configured for.

## Expected behavior

The permissions for the content type that was just deleted should be removed from all the roles they were configured for.

## Additional information

- If we decide that this is a valid request, then we should make sure that permissions are removed from roles from other places, such as:
  - when deleting text formats
  - when deleting vocabularies
  - ???

  These should be separate issues though.

- In Drupal 9.3.x with the introduction of https://www.drupal.org/node/3055548 such permissions are being cleaned up automatically, because permissions are allowed to specify dependencies.

**Comment by indigoxela** [comment]
Related tests are failing, so back to "needs work".

**Comment by klonos** [comment]
I know @indigoxela, but unfortunately I'm stuck because of this: #5354 😓

**Comment by klonos** [comment]
...so, what is the preferred way to move this forward?:
- be cheeky and call `module_invoke_all('permission')` after I create the new content type in the test
- raise a new issue, to adapt `node_type_cache_reset()` to also call `module_invoke_all('permission')` (it does make sense, since creating a new content type adds a set of 5 new permissions for it - deleting a content type should remove that set of 5 permissions)
- wait for #5354 to be accepted and merged into core
- something else??

**Comment by klonos** [comment]
@indigoxela other than the bug I've managed to surface re permissions not being refreshed when calling `backdropCreateRole()` (for which we now have a separate issue), I feel that this PR deserves 👀  so that I can address any issues or suggestions for improvement that others may have. At the same time, I don't want to upset people that might not want to see the "needs review" label added to issues unless tests for the respective PR pass.

Are you willing to "let this one slide" and allow me to set the NR label please? 🙂

**Comment by indigoxela** [comment]
> Are you willing to "let this one slide" and allow me to set the NR label please?

Things like that (issue PR blocked by other issue) happen from time to time. I think it makes sense to insist on passing test runs, as our functional tests have exactly that purpose - prevent letting something slip through, that doesn't pass testcases.

I'm aware that waiting for another PR to get merged can be frustrating, but I'd prefer to follow our usual procedures. People are able to have :eyes: on your PR here, even if it's still in state "needs work" until the blocker got merged.

**Comment by klonos** [comment]
Fair enough 🤷🏼  ...in the meantime, to get tests to pass, I've forced `checkPermissions($permissions, TRUE)` within `backdropCreateRole()` (and have added a `@todo` to remove that once we sort out #5354). This proves in practice that that's the only thing causing the tests to fail.

**Comment by stpaultim** [comment]
I was going to test this, but we don't have a working sandbox. I tried to recreate one, but failed. Also, I see that this issue may be blocked by another issue.

**Comment by klonos** [comment]
Yeah, it's been some time @stpaultim, so I'd need to rebase and see what's blocking this. Bear with me...

**Comment by klonos** [comment]
Done @stpaultim ...all tests are passing, and the PR sandbox is up and running once again 😉 

Feedback I need code-wise: should I move the code that deletes the permissions from `node_type_delete()` into a `hook_node_type_delete()` implementation in user.module, or is it fine as it is currently implemented in node.module?

**Comment by kiamlaluno** [comment]
> Feedback I need code-wise: should I move the code that deletes the permissions from `node_type_delete()` into a `hook_node_type_delete()` implementation in user.module, or is it fine as it is currently implemented in node.module?

Yes, the User module should implement `hook_node_type_delete()` to remove permissions when a content type is deleted.

**Comment by klonos** [comment]
Thought so 😅 ...also done @kiamlaluno 👍🏼

**Comment by jenlampton** [comment]
On initial review, this looks great! I am requesting one change in the code comment so that we don't accidentally make a mistake in the future :)

**Comment by robertgarrigos** [comment]
I just came across this error and see that it is almost fixed...

**Comment by robertgarrigos** [comment]
In fact, I see that some field instances configs files are not being deleted either when deleting a content type. Is there an issue for this?

**Comment by stpaultim** [comment]
@robertgarrigos or anyone else. 

This issue seems to be very close to completion, but stalled. @klonos has been busy with personal matters lately and may not have time to work on this in the near future.

I'm sure that Greg would be happy if someone recreated his PR and included the comments from @jenlampton, so that we possibly can get this one over the finish line in 1.30. 

If anyone has the time and interest.

**Comment by robertgarrigos** [comment]
I'll try it.

**Comment by robertgarrigos** [comment]
mmm I recreated the PR to add @jenlampton comment but got some coding standards failing tests:

```
Files checked: 65, Issues found: 25 in 7 files.
Error: 25 spelling issues found in 7 of the 65 files checked.
Done.
core/includes/password.inc:58:4 Unknown word (itoa)
core/includes/password.inc:58:23 Unknown word (itoa)
core/includes/password.inc:61:17 Unknown word (itoa)
core/includes/password.inc:65:17 Unknown word (itoa)
core/includes/theme.inc:1339:29 Unknown word (basetheme)
core/includes/theme.inc:1342:23 Unknown word (basetheme)
core/includes/theme.inc:1342:62 Unknown word (basetheme)
core/modules/simpletest/backdrop_web_test_case.php:1984:9 Unknown word (CURLOPT)
core/modules/simpletest/backdrop_web_test_case.php:1985:9 Unknown word (CURLOPT)
core/modules/simpletest/backdrop_web_test_case.php:1986:9 Unknown word (CURLOPT)
core/modules/simpletest/backdrop_web_test_case.php:1987:9 Unknown word (CURLOPT)
core/modules/simpletest/backdrop_web_test_case.php:1987:17 Unknown word (RETURNTRANSFER)
core/modules/simpletest/backdrop_web_test_case.php:1988:9 Unknown word (CURLOPT)
core/modules/simpletest/backdrop_web_test_case.php:2082:98 Unknown word (POSTFIELDS)
core/modules/simpletest/backdrop_web_test_case.php:23[51](https://github.com/backdrop/backdrop/actions/runs/11745111876/job/32721766581?pr=4897#step:3:52):94 Unknown word (POSTFIELDS)
```
I don't understand these errors. Why are these unknown words?

Besides, I hope I've done it right, this recreation of the PR...
